conjure-java-core/src/main/java/com/palantir/conjure/java/types/SafetyEvaluator.java

```diff
@@ -0,0 +1,285 @@
+        this.definitionMap = definitionMap;
+    }
+
+    public Optional<LogSafety> evaluate(TypeDefinition def) {
+        return Preconditions.checkNotNull(def, "TypeDefinition is required")
+                .accept(new TypeDefinitionSafetyVisitor(definitionMap, new HashSet<>()));
+    }
+
+    public Optional<LogSafety> evaluate(Type type) {
+        return Preconditions.checkNotNull(type, "TypeDefinition is required")
+                .accept(new TypeDefinitionSafetyVisitor(definitionMap, new HashSet<>()).fieldVisitor);
+    }
+
+    public Optional<LogSafety> evaluate(Type type, Optional<LogSafety> declaredSafety) {
+        return declaredSafety.or(() -> evaluate(type));
+    }
+
+    private static final class TypeDefinitionSafetyVisitor implements TypeDefinition.Visitor<Optional<LogSafety>> {
+        private final Set<TypeName> inProgress;
+        private final Type.Visitor<Optional<LogSafety>> fieldVisitor;
+
+        private TypeDefinitionSafetyVisitor(Map<TypeName, TypeDefinition> definitionMap, Set<TypeName> inProgress) {
+            this.inProgress = inProgress;
+            this.fieldVisitor = new FieldSafetyVisitor(definitionMap, this);
+        }
+
+        @Override
+        public Optional<LogSafety> visitAlias(AliasDefinition value) {
+            return with(value.getTypeName(), () -> getSafety(value.getAlias(), value.getSafety()));
+        }
+
+        @Override
+        public Optional<LogSafety> visitEnum(EnumDefinition _value) {
+            return ENUM_SAFETY;
+        }
+
+        @Override
+        public Optional<LogSafety> visitObject(ObjectDefinition value) {
+            return with(value.getTypeName(), () -> {
+                Optional<LogSafety> safety = Optional.of(LogSafety.SAFE);
+                for (FieldDefinition field : value.getFields()) {
+                    safety = combine(safety, getSafety(field.getType(), field.getSafety()));
+                }
+                return safety;
+            });
+        }
+
+        @Override
+        public Optional<LogSafety> visitUnion(UnionDefinition value) {
+            return with(value.getTypeName(), () -> {
+                Optional<LogSafety> safety = UNKNOWN_UNION_VARINT_SAFETY;
+                for (FieldDefinition variant : value.getUnion()) {
+                    safety = combine(safety, getSafety(variant.getType(), variant.getSafety()));
+                }
+                return safety;
+            });
+        }
+
+        @Override
+        public Optional<LogSafety> visitUnknown(String unknownType) {
+            throw new IllegalStateException("Unknown type: " + unknownType);
+        }
+
+        private Optional<LogSafety> with(TypeName typeName, Supplier<Optional<LogSafety>> task) {
+            if (!inProgress.add(typeName)) {
+                // Given recursive evaluation, we return the least restrictive type: SAFE.
+                return Optional.of(LogSafety.SAFE);
+            }
+            Optional<LogSafety> result = task.get();
```

Comment:
Would it make sense to cache the computed safety value for a type as this should all be statically known? But maybe not worth the effort if its fast enough already?

Reply:
Good flag regarding that edge case! Yeah, fine to leave it like this and optimize later if required!